flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/source/reader/IcebergSourceSplitReader.java

```diff
@@ -80,6 +88,35 @@ public RecordsWithSplitIds<RecordAndPosition<T>> fetch() throws IOException {
     }
 
     if (currentReader.hasNext()) {
+      if (pausedSplits.contains(currentSplitId)) {
+        // Wait until the reader is unblocked. Wake every second to catch any missed signal.
+        // Return empty records if wakeUp is called, so pauseOrResumeSplits could be processed.
+        boolean first = true;
```

Comment:
it seems that the logic can be significantly simpler if we use unlimited `wait()`. simply return an empty result upon `wakeUp.wait()`

Reply:
I am generally very-very cautious about using loop-less blocking calls.
Here is an example:
- WT (Watermark Thread)  - Watermark arrives, and the decision is to block this reader
- WT - `wakeup` is called
- ST (SplitReader Thread) - Current `fetch` finished
- ST - `pauseOrResumeSplits` is called
- ST - New `fetch` called
- ST - New `fetch` is checking if the split is paused, and find that we need to block, but do not start to wait on the lock yet.
- WT - New watermark arrives, and the decision is to unblock the reader
- WT - `wakeup` is called
- ST - If the `fetch` arrives to the locking code now, then it will not receive the `wakeup` call

The main issue is that the `wakeUp` and the `pauseOrResumeSplits` is not atomic, we need another flag to store if the given `wakeUp` is handled, or not.

So i think we minimally need to have something like this:
```
      if (pausedSplits.contains(currentSplitId)) {
          LOG.info("Reader {} paused reading split {}", indexOfSubtask, currentSplitId);
          synchronized (wakeUp) {
            if (wakeUp.get()) {
              wakeUp.set(false);
              return new RecordsBySplits(Collections.emptyMap(), Collections.emptySet());
            }

            try {
              wakeUp.wait();
            } catch (InterruptedException e) {
              Thread.currentThread().interrupt();
              throw new RuntimeException("Interrupted while blocked on reading.", e);
            }
          }
        }
      }
```

I think compared to this the added security of the `while` loop is a good compromise.
WDYT?

Reply:
I think we need the `while` loop, purely because there could be spurious wakeups and we may have to retry if `wakeUp==false`. I think the code snippet above you posted @pvary is a good compromise if you replace the `if` with a `while`.

Reply:
@mxm According to the API contract, it is ok to return empty result in case of spurious wakeups.  that is what the current PR does already.
```
            if (wakeUp.get()) {
              wakeUp.set(false);
              return new RecordsBySplits(Collections.emptyMap(), Collections.emptySet());
            }
```

Reply:
The code snippet above, and the one proposed by the PR is only different because I wanted to add logging to the blocking path, but do not want to swamp the log with a new message every second. If you think that the logging is not important, the decision is that we accept that we log a line every second on INFO level until the SplitReader is unblocked, then we can simplify the code. Or we accept that every `fetch()` loop will create and new boolean variable which is not used in 90 percent of the cases.

On the hot path I prefer optimized code to simpler code, but the difference is probably not too noticeable anyway.

Reply:
from the timeline example above, it indicates a condition where the reader thread can be unnecessary blocked for 1s, which is not ideal. trying to think if there is anyway we can avoid this suboptimal bhavior.

Reply:
The timeline is only valid if we do not use an extra flag to store the `wakeUp` state and rely on a simple `wait-notify` pattern.

I used this timeline to describe why I introduced the `wakeUp` flag, and use synchronize around it in the PR.
I make sure that the `wakeUp` flag is only read/set inside a `synchronized` block, and we return from the `fetch` while holding the synchronization lock. So if there is a concurrent `wakeUp` method call, it will be blocked until we return, and we handle the next wakeup in the next fetch loop

Reply:
> I wanted to add logging to the blocking path, but do not want to swamp the log with a new message every second.

I think it is good to log this as TRACE only. we really need an INFO for the first instance, because we already have an INFO log in `pauseOrResumeSplits` method.

it seems that we might just need this while loop
```
while (pausedSplits.contains(currentSplitId)) {
          LOG.trace("Reader {} is still paused reading split {}", indexOfSubtask, currentSplitId);
          synchronized (wakeUp) {
            if (wakeUp.get()) {
              wakeUp.set(false);
              return new RecordsBySplits(Collections.emptyMap(), Collections.emptySet());
            }

            try {
              wakeUp.wait(1000);
            } catch (InterruptedException e) {
              Thread.currentThread().interrupt();
              throw new IOException("Interrupted while paused due to watermark alignment.", e);
            }
```